File: rec-core/src/main/java/net/kimleo/rec/repository/RecRepository.java
package net.kimleo.rec.repository;

import net.kimleo.rec.Pair;
import net.kimleo.rec.repository.selector.Selector;

import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

public class RecRepository implements Iterable<RecordSet> {

    private final List<RecordSet> collections;

    private final Map<String, RecordSet> repo;

    public int getSize() {
        return collections.size();
    }

    public RecRepository(List<RecordSet> collections) {
        this.collections = collections;
        repo = toRepoMap(collections);
    }

    private Map<String, RecordSet> toRepoMap(List<RecordSet> collections) {
        return collections
                .stream()
                .map(collect -> new Pair<>(collect.getConfig().name(), collect))
                .collect(Collectors.toMap(Pair::getFirst, Pair::getSecond));
    }

    public RecordSet from(String name) {
        RecordSet set = repo.get(name);
        if (set == null) {
            throw new NoSuchElementException(String.format("Cannot found collection%s", name));
        }
        return set;
    }


    public RecRepository select(String expr) {
        return new RecRepository(Selector.of(expr).findAll(this));
    }

    public RecRepository get(String expr) {
        return select(expr);
    }

    public RecordSet get(int index) {
        return collections.get(index);
    }


    @Override
    public Iterator<RecordSet> iterator() {
        return collections.iterator();
    }
}

File: rec-core/src/main/kotlin/net/kimleo/rec/repository/RecRepository.kt
package net.kimleo.rec.repository

import net.kimleo.rec.concept.QuerySelector
import net.kimleo.rec.concept.Queryable
import net.kimleo.rec.repository.selector.Selector
import java.util.*
import java.util.function.Function

class RecRepository(val collections: List<RecordSet>): Iterable<RecordSet> {

    override fun iterator(): Iterator<RecordSet> {
        return collections.iterator()
    }

    val repo: Map<String, RecordSet> = toRepoMap(collections)
    val size = collections.size

    fun from(name: String): RecordSet = repo.getOrElse(name) {
        throw NoSuchElementException("Cannot found the collection $name.")
    }

    private fun toRepoMap(collections: List<RecordSet>): Map<String, RecordSet> {
        return collections
                .map(RecordSet::config)
                .map(RecConfig::name)
                .zip(collections).toMap()
    }

    fun select(expr: String): RecRepository {
        return RecRepository(Selector.of(expr).findAll(this))
    }

    operator fun get(expr: String): RecRepository {
        return select(expr)
    }

    operator fun get(index: Int): RecordSet {
        return collections[index]
    }
}

